### ISSUE TITLE
Track A: Enable PyTorch/MPS acceleration for RL/Neural paradigms

### ISSUE THREAD

**Issue by derrring** [derrring/MFG_PDE#111]
## Goal
Enable full PyTorch/MPS (Metal Performance Shaders) GPU acceleration on Apple Silicon (M1/M2/M3) for RL and Neural paradigms.

## Current State
- ✅ MPS detected correctly: `torch_mps: true`
- ✅ Backend initializes: `Device: mps`, `Apple Silicon: Metal Performance Shaders enabled`
- ❌ **Blocker**: Type mismatch error: `"can't assign numpy.ndarray to torch.FloatTensor"`

## Root Cause
Solvers hardcode `np.zeros()` instead of using `backend.zeros()`:

```python
# Current (broken with PyTorch backend)
def solve(self):
    U = np.zeros((Nt, Nx))  # Hardcoded NumPy
    M = np.zeros((Nt, Nx))
    # ... type error when backend is PyTorch
```

## Solution Strategy

**Fix only boundary interfaces** (5-10 files, NOT entire codebase):

1. **FixedPointIterator** - Core MFG coupling solver
2. **Solver initialization** - HJB/FP solver constructors
3. **RL environments** - PyTorch ↔ NumPy conversions at boundaries
4. **Backend propagation** - Ensure backend passed through factories

### Phase 1: Core Boundary Fixes (Week 1-2)

**File**: `mfg_pde/alg/numerical/mfg_solvers/fixed_point_iterator.py`

```python
class FixedPointIterator:
    def __init__(self, problem, hjb_solver, fp_solver, backend=None, **kwargs):
        # Ensure backend consistency
        if backend is None:
            backend = getattr(hjb_solver, 'backend', create_backend("numpy"))
        self.backend = backend
        
        # Propagate to sub-solvers
        self.hjb_solver.backend = self.backend
        self.fp_solver.backend = self.backend
    
    def solve(self, max_iterations=100, tolerance=1e-6):
        # Use backend-native array creation
        U = self.backend.zeros((Nt + 1, Nx), dtype=self.backend.float64)
        M = self.backend.zeros((Nt + 1, Nx), dtype=self.backend.float64)
        
        # Initialize with backend arrays
        U[-1, :] = self.backend.array(terminal_value)
        M[0, :] = self.backend.array(initial_density)
        # ... rest works transparently
```

**Files to modify**:
- [ ] `mfg_pde/alg/numerical/mfg_solvers/fixed_point_iterator.py`
- [ ] `mfg_pde/alg/numerical/hjb_solvers/hjb_fdm.py`
- [ ] `mfg_pde/alg/numerical/fp_solvers/fp_particle.py`
- [ ] `mfg_pde/factory.py` (solver factories)
- [ ] Other HJB/FP solver base classes as needed

### Phase 2: RL Environment Conversions (Week 3)

**Pattern**: Explicit conversions at RL ↔ Numerical boundary

```python
class MFGEnvironment(gym.Env):
    def __init__(self, problem, use_gpu=False):
        # RL: PyTorch (for GPU training)
        self.device = "mps" if use_gpu else "cpu"
        
        # Numerical solver: NumPy/JAX (fastest for CPU)
        self.solver = create_standard_solver(problem, backend="numpy")
    
    def step(self, action):
        # Convert PyTorch → NumPy
        action_np = action.detach().cpu().numpy()
        
        # Solve numerically
        reward_np = self.solver.compute_reward(action_np)
        
        # Convert NumPy → PyTorch
        reward = torch.from_numpy(reward_np).to(self.device)
        return obs, reward, done, info
```

**Files to modify**:
- [ ] `mfg_pde/alg/rl/environments/` - RL wrappers

### Phase 3: Testing & Validation (Week 4)

**New test file**: `tests/integration/test_backend_compatibility.py`

```python
@pytest.mark.parametrize("backend_name", ["numpy", "torch", "jax"])
def test_solver_backend_consistency(backend_name):
    """Verify solvers work with all backends."""
    problem = create_lq_problem()
    backend = create_backend(backend_name)
    
    solver = create_standard_solver(problem, backend=backend)
    result = solver.solve(max_iterations=10)
    
    # Should complete without type errors
    assert result.iterations == 10 or result.convergence_achieved

@pytest.mark.skipif(not torch.backends.mps.is_available(), reason="MPS not available")
def test_mps_end_to_end():
    """Test full PyTorch MPS acceleration pipeline."""
    problem = create_lq_problem()
    backend = create_backend("torch", device="mps")
    
    solver = create_standard_solver(problem, backend=backend)
    result = solver.solve(max_iterations=20, tolerance=1e-4)
    
    # Verify computation on MPS
    assert result.U.device.type == "mps"
    assert result.M.device.type == "mps"
```

**Test checklist**:
- [ ] All solvers pass with `backend="torch"`
- [ ] MPS test completes without type errors
- [ ] RL environments work with MPS backend
- [ ] Performance benchmarks (3-5x speedup expected)

## Success Metrics

**Quantitative**:
- ✅ All solvers pass with `backend="torch"`
- ✅ MPS test completes without errors
- ✅ 3-5x speedup on M1/M2/M3 for large problems (Nx > 200)

**Qualitative**:
- ✅ RL training can leverage GPU acceleration
- ✅ Neural-numerical hybrid methods work
- ✅ Examples run with `device="mps"`

## Timeline

- **Week 1-2**: Core boundary fixes (FixedPointIterator, solver init)
- **Week 3**: RL environment conversions
- **Week 4**: Testing and validation
- **Total**: ~1 month

## References

- **Proposal**: `docs/development/IMPROVEMENT_PROPOSAL_2025Q4.md`
- **Strategy**: `docs/development/future_enhancements/torch_compatibility_strategy.md`
- **Tracking**: `docs/development/future_enhancements/apple_silicon_acceleration.md`
- **Evidence**: `examples/basic/acceleration_comparison.py` (shows MPS incompatibility)

## Priority

**P0** - Critical for RL/Neural paradigm GPU acceleration

## Labels

`priority: high`, `area: algorithms`, `size: medium`, `enhancement`
